pkg/connectorrunner/runner.go

```diff
@@ -1,86 +1,158 @@
+func (c *connectorRunner) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancelCause(ctx)
+	defer cancel(ErrSigTerm)
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt)
+	go func() {
+		for range sigChan {
+			cancel(ErrSigTerm)
+		}
+	}()
 
-	err = c.manager.SaveC1Z(ctx)
+	err := c.run(ctx)
 	if err != nil {
-		logger.Error("error saving c1z", zap.Error(err))
 		return err
 	}
 
-	err = c.manager.Close(ctx)
+	err = c.Close(ctx)
 	if err != nil {
-		logger.Error("error closing c1z manager", zap.Error(err))
 		return err
 	}
 
 	return nil
 }
 
-// Run starts a connector and creates a new C1Z file.
-func (c *connectorRunner) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	defer func(c *connectorRunner, ctx context.Context) {
-		err := c.shutdown(ctx)
+func (c *connectorRunner) handleContextCancel(ctx context.Context) error {
+	l := ctxzap.Extract(ctx)
+
+	err := context.Cause(ctx)
+	if err == nil {
+		return nil
+	}
+
+	// The context was cancelled due to an expected end of the process. Swallow the error.
+	if errors.Is(err, ErrSigTerm) {
+		return nil
+	}
+
+	l.Debug("unexpected context cancellation", zap.Error(err))
+	return err
+}
+
+func (c *connectorRunner) run(ctx context.Context) error {
+	l := ctxzap.Extract(ctx)
+
+	l.Info("waiting for work....")
+
+	var done bool
+
+	for !done {
+		select {
+		case <-ctx.Done():
+			return c.handleContextCancel(ctx)
+		default:
+		}
+
+		nextTask, err := c.tasks.Next(ctx)
 		if err != nil {
-			ctxzap.Extract(ctx).Error("error shutting down", zap.Error(err))
+			l.Error("error getting next task", zap.Error(err))
+			continue
 		}
-	}(c, ctx)
 
-	sigChan := make(chan os.Signal, 1)
-	signal.Notify(sigChan, os.Interrupt)
-	go func() {
-		for range sigChan {
-			cancel()
+		// No work for us to do, go to sleep for a bit and try again
+		if nextTask == nil {
+			l.Debug("no available tasks, going to sleep for 5 seconds")
+			select {
+			case <-ctx.Done():
+				return c.handleContextCancel(ctx)
+			case <-time.After(time.Second * 5):
+			}
+			continue
 		}
-	}()
 
-	err := c.syncer.Sync(ctx)
-	if err != nil {
-		return err
+		switch nextTask.(type) {
+		case *tasks.SyncTask:
+			client, err := c.cw.C(ctx)
+			if err != nil {
+				return err
+			}
+
+			syncer, err := sdkSync.NewSyncer(ctx, client, c.dbPath)
+			if err != nil {
+				return err
+			}
+
+			err = syncer.Sync(ctx)
+			if err != nil {
+				return err
+			}
+
+			err = syncer.Close(ctx)
+			if err != nil {
+				return err
+			}
+
+			err = c.tasks.Finish(ctx, nextTask.GetTaskId())
+			if err != nil {
+				return err
+			}
+
+			if c.onDemandMode {
+				done = true
+			}
+
+			err = c.cw.Close()
+			if err != nil {
+				return err
+			}
+
+		default:
+			l.Debug("unknown task type, going to sleep for 10 seconds")
+			select {
+			case <-ctx.Done():
+				return c.handleContextCancel(ctx)
+			case <-time.After(time.Second * 10):
```

Comment:
Another randomly picked duration to wait to avoid spinning cores. I assume we want to be resilient to unknown things, so I don't exit here.